lib/sinon/spy.js

```diff
@@ -53,9 +53,11 @@ function matchingFake(fakes, args, strict) {
         return undefined;
     }
 
-    for (var i = 0, l = fakes.length; i < l; i++) {
-        if (fakes[i].matches(args, strict)) {
-            return fakes[i];
+    var fakesIndex = fakes.length;
+
+    while (fakesIndex--) {
```

Comment:
We don't yet have style and compatibility guidelines for the effort of creating v2.x. I'll try to get that discussion going in a separate thread.

For this PR, I would suggest using [`Array.prototype.filter`][filter] instead of a while loop. `filter` is defined in ES5.1, and we're already using `forEach`, so we have committed ourselves it ES5.1 by now.

```javascript
function matchingFake(fakes, args, strict) {
    if (!fakes || fakes.length === 0) {
        return undefined;
    }

    var matchingFakes = fakes.filter(function (f) {
        return f.matches(args, strict);
    });

    return matchingFakes.pop() || undefined;
}
```

I am aware that the solution iterates over ALL the fakes ... but there's not likely to be so many that we can actually measure the difference under anything other than truly extraordinary circumstances.

I think this refactoring of your changes works (all the tests pass). What do you think?

[filter]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter